### PR TITLE
[Perf][VoxCPM2] Add unified decode graph optimization

### PR DESCRIPTION
**Pull request by Sy0307**
PLEASE FILL IN THE PR DESCRIPTION HERE.

## Purpose

This PR improves VoxCPM2 TTS serving latency by reducing Python routing and kernel launch overhead in the decode path.

VoxCPM2 decode currently crosses multiple small stages per step, including scaffold decode, FSQ/fusion, residual decode, CFM solve, feature encoding, stop prediction, and audio collection. The segmented path leaves overhead between these stages, especially under high concurrency.

This PR adds CUDA graph support for the VoxCPM2 decode path while keeping the default behavior quality-safe.

Key changes:
- Add decode tail graph support for the CFM / feature encoder / stop prediction tail.
- Add an opt-in full unified decode graph path for scaffold, residual, CFM, feature encoding, and audio-side state commit.
- Add runner-assisted attention metadata support so full decode graph capture can reuse vLLM-built paged-attention metadata instead of reconstructing it inside the model layer.
- Add bounded graph bucketing for dynamic decode batch sizes, avoiding all-size graph capture and reducing fallback for non-power-of-two active batches.
- Preserve VoxCPM2 runtime configuration through the existing deploy YAML / engine overrides path.
- Add static regression coverage for the runner-assisted graph hooks and bucketed unified graph path.

## Test Plan

- [x] `ruff check .`
- [x] `ruff format --check .`
- [x] `python3 -m py_compile vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py vllm_omni/worker/gpu_ar_model_runner.py tests/model_executor/models/voxcpm2/test_runner_assisted_unified_graph_static.py`
- [x] `python3 tests/model_executor/models/voxcpm2/test_runner_assisted_unified_graph_static.py`
- [x] Remote H20 c=64 VoxCPM2 HTTP benchmark
- [x] Verified output bytes/request stayed in the expected range, with 100% success and no stop runaway

**vLLM Version:**

0.23.0

**vLLM-Omni Commit:**

d295fcdb028b9070de5e150119f0a1da5a6e33d9

## Test Result

H20 benchmark, VoxCPM2 HTTP serving, `c=64`.

| Metric | Baseline segmented | Unified graph fixed | Delta |
| --- | ---: | ---: | ---: |
| TTFB mean | 1.945s | 1.06-1.10s | ~44-46% lower |
| Wall mean | 8.26s | 4.56s | ~45% lower |
| Throughput | 6.4 req/s | 11.7-12.1 req/s | up to ~89% higher |
| Bytes/request | 498-517 KB | ~526 KB | expected range |
| Success | 100% | 100% | no regression observed |

Additional validation:
- Warm/capture run completed with `64/64 success`, `0 errors`, `Bytes mean=525.9KB`.
- Steady run 1: `TTFB mean=1.101s`, `Wall mean=4.566s`, `Throughput=11.7 req/s`, `Bytes mean=525.5KB`.
- Steady run 2: `TTFB mean=1.060s`, `Wall mean=4.563s`, `Throughput=12.1 req/s`, `Bytes mean=527.8KB`.

Runner-assisted bucket logs confirmed non-power-of-two active batches are padded into bounded graph buckets, for example:
- `num_reqs=47 padded_reqs=48`
- `num_reqs=23 padded_reqs=24`
- `num_reqs=55 padded_reqs=56`

This avoids the previous exact-size/power-of-two fallback pattern without pre-capturing every batch size.

PTAK @linyueqian @hsliuustc0106

### CODE REVIEW

**Review by hsliuustc0106 — COMMENTED** [review]
Performance optimization looks well-documented with solid benchmark evidence. All tests pass.

File: vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py [hunk 1]
```diff
@@ -1536,8 +1751,508 @@ def _run_cfm_cuda_graph_to_state_buffer(
...
+            return next_embed, stop_out, cfm_out
+
+        original_estimator = tts.feat_decoder.estimator
+        if not hasattr(self, "_estimator_for_unified_capture"):
+            estimator_for_compile = (
+                tts.feat_decoder.estimator._orig_mod
+                if hasattr(tts.feat_decoder.estimator, "_orig_mod")
+                else original_estimator
+            )
+            self._estimator_for_unified_capture = torch.compile(
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py:1962]: Please make this graph-capture path fall back if compile/capture fails. `enable_decode_tail_graph` is enabled in `voxcpm2.yaml`, but this re-runs `torch.compile` after `_setup_torch_compile()` may already have logged a compile failure, so an exception here fails the request instead of using `_forward_decode_tail_graph_fallback`.

File: vllm_omni/worker/gpu_ar_model_runner.py [hunk 2]
```diff
@@ -395,6 +395,147 @@ def _model_needs_full_prefix_hidden_states(self) -> bool:
         model = getattr(self, "model", None)
         return bool(getattr(model, "requires_full_prefix_cached_hidden_states", True))
 
+    def _needs_runner_assisted_full_attention_metadata(
```
Comment (hsliuustc0106) [inline vllm_omni/worker/gpu_ar_model_runner.py]: do these changes only apply to voxcpm2 now? how other models may benefit from the changes to modelrunner

File: vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py [hunk 3]
```diff
@@ -1106,6 +1189,28 @@ def _setup_cfm_buffers(self) -> None:
             max_batch_size=self._max_batch_size,
         )
 
+    @staticmethod
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py:1192]: the file change looks not model specific

**Comment by hsliuustc0106** [comment]
I think it's better to 

1. provide kernel execution time w/o this commit.
2. think of better extraction for changes to modelrunner and model files, not sure whether this can apply to other models as well
3. make minimal changes to modelrunner which is very hard to maintain

File: vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py [hunk 4]
```diff
@@ -1536,8 +1751,508 @@ def _run_cfm_cuda_graph_to_state_buffer(
...
+        inputs_embeds: torch.Tensor,
+        positions: torch.Tensor,
+        num_reqs: int,
+    ) -> torch.Tensor:
+        graph_size = self._select_unified_graph_bucket_size(num_reqs)
+        if graph_size is None:
+            raise RuntimeError(f"No unified decode graph bucket for batch size {num_reqs}")
+        g = self._unified_graphs.get(graph_size)
+        if g is None:
+            g = self._capture_unified_decode_graph(graph_size)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py:2095]: Please give the lazy unified graph capture the same failure containment as the fallback path. Since `enable_unified_decode_graph` is enabled in the default deploy config, any compile/capture exception here fails the request instead of falling back to the segmented decode path. There is already a thread for the tail graph capture case; this is the equivalent unified path.

File: vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py [hunk 5]
```diff
@@ -1352,6 +1495,78 @@ def _should_use_decode_graph(self, batch_size: int) -> bool:
...
+                try:
+                    candidates.append(int(part))
+                except ValueError:
+                    logger.warning("Ignoring invalid graph pre-capture size: %s", part)
+
+        sizes = {
+            size
+            for size in candidates
+            if 0 < size <= max_size
+            and size in self._unified_graph_bucket_sizes
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py:1552]: This shared parser filters every pre-capture size through `_unified_graph_bucket_sizes`, even for decode-tail graphs. With the PR's default `unified_decode_graph_max_batch_size: 1`, `decode_tail_graph_pre_capture_sizes: power2` collapses to only size 1, so batch 2/4/8/... tail graphs are not pre-captured. Decode-tail sizes should use a tail-specific bucket/candidate set, or skip the unified bucket filter here.

File: vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py [hunk 6]
```diff
@@ -1536,8 +1751,508 @@ def _run_cfm_cuda_graph_to_state_buffer(
...
+        def unified_fwd():
+            scaffold_h = self.model(input_ids=None, positions=g.positions, inputs_embeds=g.input_embeds)
+            if isinstance(scaffold_h, tuple):
+                scaffold_h = scaffold_h[0]
+            lm_h = tts.fsq_layer(scaffold_h)
+            res_input = tts.fusion_concat_proj(torch.cat([lm_h, g.prev_feat_embed], dim=-1))
+            res_h = self.residual_model(positions=g.positions, inputs_embeds=res_input)
+            dit_h = dit_proj(lm_h, res_h)
+            cond = g.prefix_feat_cond.transpose(1, 2).contiguous()
+            g.cfm_noise.normal_()
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/voxcpm2/voxcpm2_talker.py:1805]: This bypasses `deterministic_cfm_noise`: the new unified graph records random noise generation inside the graph body, while the existing CFM/fallback paths fill per-request noise with `_fill_deterministic_cfm_noise` before replay. The tail graph has the same pattern at `tail_fwd()`. Please either make noise an explicit static input buffer copied before replay, or disable these graph paths when deterministic mode is requested.

**Review by hsliuustc0106 — COMMENTED** [review]
Static review follow-up: a few graph-path issues that look worth fixing before relying on the new defaults.

**Review by hsliuustc0106 — COMMENTED** [review]
I’m concerned that this PR moves a large amount of model-specific CUDA graph coordination into `GPUARModelRunner`, which is a core runner shared by all AR models. The benchmark and validation are VoxCPM2-only, but the runner hooks, padding behavior, attention metadata changes, and graph-capture mode changes affect the common execution path and will become maintenance burden for future models.

Could we keep the runner change smaller and more generic? For example:
- expose a minimal model-declared graph/metadata capability interface from the runner;
- keep VoxCPM2-specific bucketing, graph lifecycle, capture policy, and fallback logic inside the VoxCPM2 model/talker layer;
- add runner-level tests with a small fake model to prove the generic hooks are safe for non-VoxCPM2 models;
- add explicit documentation of the hook contract, especially around padded req counts, attention metadata, slot mappings, fallback behavior, and capture failure handling.

Without that separation, future model-runner changes will need to understand VoxCPM2 graph internals, and future models may accidentally inherit behavior that was only validated for VoxCPM2.
